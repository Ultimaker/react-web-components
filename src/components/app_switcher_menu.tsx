import * as React from 'react';
import classNames from 'classnames';

// components
import DropDownMenuBase from './drop_down_menu_base';
import Button from './button';
import Divider from './divider';
import AppSwitcherTrigger from './app_switcher_trigger';
import SlideInPanel from '../components/slide_in_panel';
import LinkIcon from './icons/link_icon';

export interface ApplicationListItem {
    /* eslint-disable camelcase */
    application_type?: string,
    name?: string,
    url?: string,
}

export interface AppSwitcherMenuProps {
    /** The list of apps to show */
    appsList: ApplicationListItem[];
    /**
     * Used to toggle the visibility of the menu.
     * Passed in by the header component.
     */
    showMenu?: boolean;
    /**
     * Called when the menu visibility is toggled.
     * Passed in by the header component.
     */
    onToggleMenuHandler?: (showMenu: boolean) => void;
}

export class AppSwitcherMenu extends React.Component<AppSwitcherMenuProps, {}> {
    private _onToggleMenuHandler(showMenu: boolean) {
        const { onToggleMenuHandler } = this.props;
        onToggleMenuHandler(showMenu);
    }

    private _renderLink(link: ApplicationListItem): JSX.Element {
        return (
            <li key={link.url}>
                <Button
                    appearance="no-style"
                    className="drop-down-menu-base__item"
                    type="link"
                    linkURL={link.url}
                    linkToNewTab={link.application_type === 'external'}
                    onClickHandler={() => this._onToggleMenuHandler(false)}
                >
                    {link.name}
                    {link.application_type === 'external'
                        && <LinkIcon />}
                </Button>
            </li>
        );
    }

    private _renderInternalLinks(): JSX.Element[] {
        const { appsList } = this.props;
        const internalLinks = appsList.filter((app) => app.application_type === 'internal');
        return internalLinks.map((link) => this._renderLink(link));
    }

    private _renderExternalLinks(): JSX.Element[] {
        const { appsList } = this.props;
        const internalLinks = appsList.filter((app) => app.application_type === 'external');
        return internalLinks.map((link) => this._renderLink(link));
    }

    render(): JSX.Element {
        const { showMenu } = this.props;
        const { appsList } = this.props;

        return (
            <div className={classNames('app-switcher-menu', { 'app-switcher-menu--open': showMenu })}>
                <div className="app-switcher-menu__desktop show-sm">
                    <DropDownMenuBase
                        showMenu={showMenu}
                        triggerElement={<AppSwitcherTrigger isAppSwitcherOpen={showMenu} />}
                        onToggleMenuHandler={(newShowMenu) => this._onToggleMenuHandler(newShowMenu)}
                    >
                        {appsList && (
                            <div>
                                {this._renderInternalLinks()}

                                <Divider width="1px" margin />

                                {this._renderExternalLinks()}
                            </div>
                        )}
                    </DropDownMenuBase>
                </div>

                <div className="app-switcher-menu__mobile hide-sm">
                    <Button
                        appearance="no-style"
                        onClickHandler={() => this._onToggleMenuHandler(!showMenu)}
                        className="app-switcher-menu__mobile-trigger"
                    >
                        <AppSwitcherTrigger isAppSwitcherOpen={showMenu} />
                    </Button>

                    <SlideInPanel
                        isOpen={showMenu}
                        width="30rem"
                        includeHeader={false}
                        onOverlayClickHandler={() => this._onToggleMenuHandler(false)}
                    >

                        <ul className="drop-down-menu-base__menu-list">
                            {appsList && (
                                <div>
                                    {this._renderInternalLinks()}

                                    <Divider width="1px" margin />

                                    {this._renderExternalLinks()}
                                </div>
                            )}
                        </ul>
                    </SlideInPanel>
                </div>
            </div>
        );
    }
}

export default AppSwitcherMenu;
