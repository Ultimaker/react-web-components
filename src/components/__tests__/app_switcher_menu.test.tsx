// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import { shallow } from 'enzyme';

// component
import { AppSwitcherMenu, AppSwitcherMenuProps } from '../app_switcher_menu';

describe('The AppSwitcherMenu component', () => {
    let props: AppSwitcherMenuProps;
    let wrapper;

    beforeEach(() => {
        props = {
            appsList: [
                {
                    application_type: 'internal',
                    name: 'My Cloud Portal',
                    url: 'https://mycloud-staging.ultimaker.com',
                },
                {
                    application_type: 'internal',
                    name: 'Account',
                    url: 'https://account-staging.ultimaker.com',
                },
                {
                    application_type: 'internal',
                    name: 'Marketplace Portal',
                    url: 'https://contribute-staging.ultimaker.com',
                },
                {
                    application_type: 'external',
                    name: 'Ultimaker Community',
                    url: 'https://community.ultimaker.com',
                },
                {
                    application_type: 'external',
                    name: 'Ultimaker.com',
                    url: 'https://ultimaker.com',
                },
            ],
        };
        wrapper = shallow(<AppSwitcherMenu {...props} />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
