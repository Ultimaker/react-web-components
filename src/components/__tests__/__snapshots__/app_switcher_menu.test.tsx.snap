// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AppSwitcherMenu component should render 1`] = `
<div
  className="app-switcher-menu"
>
  <div
    className="app-switcher-menu__desktop show-sm"
  >
    <DropDownMenuBase
      menuDirection="south"
      onToggleMenuHandler={[Function]}
      triggerElement={<AppSwitcherTrigger />}
    >
      <div>
        <li
          key="https://mycloud-staging.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={false}
            linkURL="https://mycloud-staging.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            My Cloud Portal
          </Button>
        </li>
        <li
          key="https://account-staging.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={false}
            linkURL="https://account-staging.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Account
          </Button>
        </li>
        <li
          key="https://contribute-staging.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={false}
            linkURL="https://contribute-staging.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Marketplace Portal
          </Button>
        </li>
        <Divider
          direction="horizontal"
          margin={true}
          width="1px"
        />
        <li
          key="https://community.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={true}
            linkURL="https://community.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Ultimaker Community
            <Icon(LinkIcon) />
          </Button>
        </li>
        <li
          key="https://ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={true}
            linkURL="https://ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Ultimaker.com
            <Icon(LinkIcon) />
          </Button>
        </li>
      </div>
    </DropDownMenuBase>
  </div>
  <div
    className="app-switcher-menu__mobile hide-sm"
  >
    <Button
      appearance="no-style"
      className=""
      linkToNewTab={false}
      onClickHandler={[Function]}
      shape="rectangle"
      type="button"
    >
      <AppSwitcherTrigger />
    </Button>
    <SlideInPanel
      includeHeader={false}
      onOverlayClickHandler={[Function]}
      slideDirection="right"
      width="30rem"
    >
      <ul
        className="drop-down-menu-base__menu-list"
      >
        <li
          key="https://mycloud-staging.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={false}
            linkURL="https://mycloud-staging.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            My Cloud Portal
          </Button>
        </li>
        <li
          key="https://account-staging.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={false}
            linkURL="https://account-staging.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Account
          </Button>
        </li>
        <li
          key="https://contribute-staging.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={false}
            linkURL="https://contribute-staging.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Marketplace Portal
          </Button>
        </li>
        <Divider
          direction="horizontal"
          margin={true}
          width="1px"
        />
        <li
          key="https://community.ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={true}
            linkURL="https://community.ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Ultimaker Community
            <Icon(LinkIcon) />
          </Button>
        </li>
        <li
          key="https://ultimaker.com"
        >
          <Button
            appearance="no-style"
            className="drop-down-menu-base__item"
            linkToNewTab={true}
            linkURL="https://ultimaker.com"
            onClickHandler={[Function]}
            shape="rectangle"
            type="link"
          >
            Ultimaker.com
            <Icon(LinkIcon) />
          </Button>
        </li>
      </ul>
    </SlideInPanel>
  </div>
</div>
`;
